src/main/java/org/quiltmc/loader/impl/transformer/ChasmInvoker.java

```diff
@@ -0,0 +1,243 @@
+				continue;
+			}
+			Files.walkFileTree(path2, new SimpleFileVisitor<Path>() {
+				@Override
+				public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+					if (file.getFileName().toString().endsWith(".class")) {
+						package2mod.put(path2.relativize(file.getParent()).toString(), mod.id());
+						inputClassCache.put(path2.relativize(file).toString(), Files.readAllBytes(file));
+					}
+					return FileVisitResult.CONTINUE;
+				}
+			});
+		}
+
+		ChasmProcessor chasm = new ChasmProcessor(new Context() {
+
+			@Override
+			public byte @Nullable [] readFile(String path) {
+
+				byte[] cached = inputClassCache.get(path);
+				if (cached != null) {
+					return cached;
+				}
+
+				// Chasm wants this to be pure, which it is!
+				// (since the transform cache isn't modified during chasm invocation anyway)
+
+				for (ModLoadOption mod : modList) {
+					Path path2 = root.resolve(mod.id()).resolve(path);
+					if (Files.isRegularFile(path2)) {
+						try {
+							return Files.readAllBytes(path2);
+						} catch (IOException e) {
+							return null;
+						}
+					}
+				}
+				return null;
+			}
+
+			@Override
+			public boolean isInterface(String className) {
+				ClassReader cr = new ClassReader(readFile(LoaderUtil.getClassFileName(className)));
+				ClassInfoGrabber namer = new ClassInfoGrabber();
+				cr.accept(namer, ClassReader.SKIP_CODE | ClassReader.SKIP_DEBUG | ClassReader.SKIP_FRAMES);
+				return (Opcodes.ACC_INTERFACE & namer.access) != 0;
+			}
+
+			@Override
+			public boolean isAssignable(String leftClass, String rightClass) {
+				if ("java/lang/Object".equals(leftClass)) {
+					return true;
+				}
+
+				byte[] file = readFile(LoaderUtil.getClassFileName(rightClass));
+				if (file == null) {
+					return false;
+				}
+
+				ClassReader cr = new ClassReader(file);
+				ClassInfoGrabber namer = new ClassInfoGrabber();
+				cr.accept(namer, ClassReader.SKIP_CODE | ClassReader.SKIP_DEBUG | ClassReader.SKIP_FRAMES);
+
+				if (namer.superName.equals(leftClass)) {
+					return true;
+				}
+
+				for (String itf : namer.interfaces) {
+					if (itf.equals(leftClass)) {
+						return true;
+					}
+				}
+
+				if (isAssignable(leftClass, namer.superName)) {
+					return true;
+				}
+
+				for (String itf : namer.interfaces) {
+					if (isAssignable(leftClass, itf)) {
+						return true;
+					}
+				}
+
+				return false;
+			}
+
+			@Override
+			public String getSuperClass(String className) {
+				ClassReader cr = new ClassReader(readFile(LoaderUtil.getClassFileName(className)));
+				ClassInfoGrabber namer = new ClassInfoGrabber();
+				cr.accept(namer, ClassReader.SKIP_CODE | ClassReader.SKIP_DEBUG | ClassReader.SKIP_FRAMES);
+				return namer.superName;
+			}
+		});
+
+		for (ModLoadOption mod : modList) {
+			Path modPath = root.resolve(mod.id());
+			if (!Files.exists(modPath)) {
+				continue;
+			}
+			Path chasmRoot = modPath.resolve("chasm_transformers");
```

Comment:
Ah, I didn't think about this in terms of URLs & domains - it makes a lot of sense to use that instead.

Reply:
As mentioned in discord, another option would be to have the root path be user-defined in QMJ, since people might find it strange to use someone elses namespace.

Reply:
Yeah, that's definitely a sensible option - but I'd rather not touch QMJ when we're still in the experimental stage.

Reply:
Okay, I take that back - now chasm files are expected to be loaded from a folder specified by `experimental_chasm_transformers` in the root of the `quilt.mod.json`.

Reply:
It seems like other people preferred this as well, so sounds good.